pkg/transform/transform.go

```diff
@@ -0,0 +1,158 @@
+package transform
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/fusor/cpma/env"
+	"github.com/fusor/cpma/pkg/io"
+	"github.com/fusor/cpma/pkg/transform/oauth"
+	"github.com/fusor/cpma/pkg/transform/secrets"
+	"github.com/sirupsen/logrus"
+)
+
+// OCP4InstallMsg message about using generated manifests
+const OCP4InstallMsg = `To install OCP4 run the installer as follow in order to add CRDs:
+' /openshift-install --dir $INSTALL_DIR create install-config'
+'./openshift-install --dir $INSTALL_DIR create manifests'
+# Copy generated CRD manifest files  to '$INSTALL_DIR/openshift/'
+# Edit them if needed, then run installation:
+'./openshift-install --dir $INSTALL_DIR  create cluster'`
+const MasterConfigFile = "/etc/origin/master/master-config.yaml"
+const NodeConfigFile = "/etc/origin/node/node-config.yaml"
+const RegistriesConfigFile = "/etc/containers/registries.conf"
+
+type Cluster struct {
+	Master Master
+}
+
+type Master struct {
+	OAuth   oauth.OAuthCRD
+	Secrets []secrets.Secret
+}
+
+type Manifest struct {
+	Name string
+	CRD  []byte
+}
+
+type Config struct {
+	MasterConfigFile     string
+	NodeConfigFile       string
+	RegistriesConfigFile string
+	OutputDir            string
+	Hostname             string
+}
+
+type ManifestTransformOutput struct {
+	//Config    Config
+	Manifests []Manifest
+}
+
+type TransformRunner struct {
+	Config string
+}
+
+// GetFile allows to mock file retrieval
+var GetFile = io.GetFile
+
+func Start() {
+	config := Config{}
+	config.OutputDir = env.Config().GetString("OutputDir")
+	config.Hostname = env.Config().GetString("Source")
+	config.MasterConfigFile = MasterConfigFile
+	config.RegistriesConfigFile = RegistriesConfigFile
+	transformRunner := NewTransformRunner(config)
+
+	if err := transformRunner.Transform([]Transform{
+		OAuthTransform{
+			Config: &config,
+		},
+		SDNTransform{
+			Config: &config,
+		},
+		RegistriesTransform{
+			Config: &config,
+		},
+	}); err != nil {
+		logrus.WithError(err).Fatalf("%s", err.Error())
+	}
+}
+
+// DumpManifests creates OCDs files
+func DumpManifests(manifests []Manifest) {
+	for _, manifest := range manifests {
+		maniftestfile := filepath.Join(env.Config().GetString("OutputDir"), "manifests", manifest.Name)
+		os.MkdirAll(path.Dir(maniftestfile), 0755)
+		err := ioutil.WriteFile(maniftestfile, manifest.CRD, 0644)
+		logrus.Printf("CRD:Added: %s", maniftestfile)
+		if err != nil {
+			logrus.Panic(err)
+		}
+	}
+}
+
+func (config *Config) Fetch(path string) []byte {
+	dst := filepath.Join(config.OutputDir, config.Hostname, path)
+	f := GetFile(config.Hostname, path, dst)
+	logrus.Printf("File:Loaded: %s", dst)
+
+	return f
+}
+
+type Extraction interface {
+	Transform() (TransformOutput, error)
+	Validate() error
+}
+
+type Transform interface {
```

Comment:
I originally had called this a transform, but I've been uncertain of the correct language here that sufficiently describes the concept. @jwmatthews you've called this a `Component` in a lot of our conversations. What I'd like for this interface to represent is one of the encapsulated features to be moved over from ocp3 to ocp4. I settled on calling this a `Migration` on the most recent version of my poc, which represents the flow of actions that happen for a given feature. Part of that is the ETL, but we already identified another possible step we want to include that is "inspect the extraction and determine if the customer has made changes, and apply a <green>, <yellow>, or <red> status" to the incoming data. I don't have a full grasp of how this work work in my mind, but I think this is the natural place to possible put that step.

Reply:
Not knowing the full scope, `Transform` seems okay for now. If it changes to `Migration` that's probably okay too.